proposals/companion-vals.md

```diff
@@ -0,0 +1,198 @@
+```kotlin
+class A {
+  private companion val logger: Logger
+     get() = Logger.getLogger("LogName")
+  
+  fun foo() = info("Log info") //logger.info("Log info")
+}
+```
+
+### Mutable companion object
+Ability to declare mutable companion value
+```kotlin
+class A {
+  private companion var logger: Logger = Logger.getLogger("LogName")
+  
+  fun foo() = info("Log info") //logger.info("Log info")
+}
+```
+**lateinit** and **by** modifiers are also supported.
+
+### Declaration at parameters level
+Ability to declare companion value as constructor parameter.
+
+```kotlin
+class A(private companion val input: B()){
+  fun foo() = bar() //B.bar()
+}
+```
+This kind of declaration supports only named companion value.
+
+### Declaration inside of function signature
+Implicit access to object members passed as parameter
+```kotlin
+fun foo(companion b: Bar, x: Int, y: Int){
+  baz() //the same as b.baz()
+}
+```
+
+It is possible to declare two companion parameters but with different types. The following code is illegal:
+```kotlin
+fun foo(companion x: Int, companion y: Int){
+  toLong() //which implicit receiver we should use? x or y?
+}
+```
+
+## Compiler implementation
+Compiler implementation is trivial: every composed object should be transformed into `private val` (or `protected val`, according with composition visibility modifier) declaration:
+
+Given:
+```kotlin
+class Derived(a: Int, b: Int): Base2(a), Serializable{
+  private companion val base1 = Base1(a + b)
+
+  fun foo() = baz() //baz() from Base1
+}
+```
+
+Transformation result:
+```kotlin
+class Derived(a: Int, b: Int): Base2(a), Serializable {
+  private val base2 = Base1(a + b)
+
+  fun foo() = base2.baz()
+}
+```
+
+For companion values without explicitly defined name compiler should emit **synthetic** modifier.
+
+Given:
+```kotlin
+open class Derived(a: Int, b: Int): Base2(a), Serializable{
+  protected companion val: Base1(a + b)
+
+  fun foo() = baz() //baz() from Base1
+}
+```
+
+Transformation result:
+```kotlin
+open class Derived(a: Int, b: Int): Base2(a), Serializable {
+  @JvmSynthetic
+  protected val $Base1_companion$ = Base1(a + b)
+
+  fun foo() = $Base1_companion$.baz()
+}
+```
+
+The complexity of compilation lies in the field of lexical scope resolution for the specified name. To solve this issue, the following resolution order can be applied:
+1. Looking for own member in the top-level class
+1. Looking for member in the inherited class
+1. Looking for member in companion object (if declared)
+1. Looking for member in implicitly composed/aggregated object
+
+### Limitations
+Companion value cannot have nullable type.
+Two companion values cannot have identical types.
+
+## Practical use
+Logging methods can be accessible implictly inside of class:
+```kotlin
+open class A{
+  protected companion val: ILogger = LoggerFactory.getLogger(A::class.qualifiedName)
+  
+  fun foo(){
+    info("Log entry") //the same as this@ILogger.info
```

Comment:
Why not just `logger.info()`? To save a few characters?

Reply:
`companion object` also helps to save few characters as well as `apply` and `run` functions from stdlib. The same story here.

Reply:
But one big difference between this feature and apply/run/let/etc that scope functions do not require any additional language features, all of them are just library functions

Reply:
Agree, but companion object is a language feature with similar behavior.